stdlib/public/core/UnicodeHelpers.swift

```diff
@@ -84,6 +84,22 @@ internal func _decodeUTF8(
   return Unicode.Scalar(_unchecked: value)
 }
 
+internal func _decodeScalar(
+  _ utf16: UnsafeBufferPointer<UInt16>, startingAt i: Int
+) -> (Unicode.Scalar, scalarLength: Int) {
+  let high = utf16[i]
+  if i + 1 >= utf16.count {
+    return (Unicode.Scalar(_unchecked: UInt32(high)), 1)
+  }
+  
+  if !_isLeadingSurrogate(high) {
+    return (Unicode.Scalar(_unchecked: UInt32(high)), 1)
+  }
+  
+  let low = utf16[i+1]
+  return (Unicode.Scalar(_unchecked: _decodeSurrogatePair(leading: high, trailing: low)), 2)
+}
```

Comment:
Add an `_internalInvariant` check to each of these code paths that any surrogate is properly paired.